include/deal.II/matrix_free/evaluation_kernels_face.h

```diff
@@ -1667,129 +1684,263 @@ namespace internal
+    static bool
+    evaluate_tensor(const unsigned int                     n_components,
+                    const EvaluationFlags::EvaluationFlags evaluation_flag,
+                    const Number                          *values_dofs,
+                    FEEvaluationData<dim, Number, true>   &fe_eval)
+    {
+      const auto &shape_info = fe_eval.get_shape_info();
+      const auto &shape_data = shape_info.data.front();
+
+      const unsigned int dofs_per_comp_face =
+        fe_degree > -1 ?
+          Utilities::pow(fe_degree + 1, dim - 1) :
+          Utilities::fixed_power<dim - 1>(shape_data.fe_degree + 1);
+
+      // Note: we always keep storage of values, 1st and 2nd derivatives in an
+      // array, so reserve space for all three here
+      Number *temp         = fe_eval.get_scratch_data().begin();
+      Number *scratch_data = temp + 3 * n_components * dofs_per_comp_face;
+
+      bool use_vectorization = true;
+      if (fe_eval.get_dof_access_index() ==
+            MatrixFreeFunctions::DoFInfo::dof_access_cell &&
+          fe_eval.is_interior_face() == false) // exterior faces in the ECL loop
+        for (unsigned int v = 0; v < Number::size(); ++v)
+          if (fe_eval.get_cell_ids()[v] != numbers::invalid_unsigned_int &&
+              fe_eval.get_face_no(v) != fe_eval.get_face_no(0))
+            use_vectorization = false;
+
+      project_to_face<fe_degree>(n_components,
+                                 evaluation_flag,
+                                 values_dofs,
+                                 fe_eval,
+                                 use_vectorization,
+                                 temp,
+                                 scratch_data);
+
+      evaluate_in_face<fe_degree, n_q_points_1d>(
+        n_components, evaluation_flag, fe_eval, temp, scratch_data);
+
+      if (dim == 3)
+        adjust_quadrature_for_face_orientation(
+          n_components, evaluation_flag, fe_eval, use_vectorization, temp);
 
       return false;
     }
+
+    template <int fe_degree, int n_q_points_1d>
+    static bool
+    run(const unsigned int                     n_components,
+        const EvaluationFlags::EvaluationFlags evaluation_flag,
+        const Number                          *values_dofs,
+        FEEvaluationData<dim, Number, true>   &fe_eval)
+    {
+      const auto &shape_info = fe_eval.get_shape_info();
+
+      if (shape_info.element_type == MatrixFreeFunctions::tensor_none)
+        return evaluate_tensor_none(n_components,
+                                    evaluation_flag,
+                                    values_dofs,
+                                    fe_eval);
+      else
+        return evaluate_tensor<fe_degree, n_q_points_1d>(n_components,
+                                                         evaluation_flag,
+                                                         values_dofs,
+                                                         fe_eval);
```

Comment:
Why is the split up needed?

Reply:
The code is much clearer to me this way concerning templates and control flow.